### PR TITLE
arkouda_server refactor

### PR DESCRIPTION
**Pull request by hokiegeek2**
This is a draft PR that corresponds to a proposal to foster integration of Arkouda server implementations by doing the following:

1. refactors arkouda_server to be a driver, moving server daemon logic to the ServerDaemon module
2. adds an ServerDaemon module that encapsulates 1..n ArkoudaServerDaemon implementations
3. adds a getServerDaemon factory that returns an ArkoudaServerDaemon implementation per a CLI arg

### CODE REVIEW

**Review by Ethan-DeBandi99 — COMMENTED** [review]
I think this refactor looks good. I don't see anything that jumps out as an issue.

File: src/ServerDaemon.chpl [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+    /**
+     * The BaseServerDaemon class serves as the base Arkouda server
+     * daemon which is run within the arkouda_server driver
+     */
+    class BaseServerDaemon : ArkoudaServerDaemon {
+        var serverToken : string;
+        var arkDirectory : string;
+        var serverMessage : string;
+        var reqCount: int = 0;
+        var repCount: int = 0;
```
Comment (pierce314159) [inline src/ServerDaemon.chpl:71]: this is super nitpicky but i don't love how similar these var names are. It's not a problem but seems easy to get mixed up

Comment (pierce314159) [reply src/ServerDaemon.chpl:71]: Also just realized these names are already in the server so yeah there's no need to change this

File: src/ServerDaemon.chpl [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+            // intercepted in the cmd processing select statement and processed specially
+            registerFunction("array", akMsgSign);
+            registerFunction("connect", akMsgSign);
+            registerFunction("disconnect", akMsgSign);
+            registerFunction("noop", akMsgSign);
+            registerFunction("ruok", akMsgSign);
+            registerFunction("shutdown", akMsgSign);
+
+            // Add the dynamic Modules/cmds implemented via ServerRegistration.chpl & ServerModules.cfg
+            doRegister();
```
Comment (pierce314159) [inline src/ServerDaemon.chpl:284]: I know this was put up before the change but just wanted to note that this is no longer needed as of #1606

**Review by pierce314159 — APPROVED** [review]
Looks good!

**Review by joshmarshall1 — APPROVED** [review]

**Comment by hokiegeek2** [comment]
@Ethan-DeBandi99 @pierce314159 @joshmarshall1 @jeichert60 thanks for your reviews!

**Comment by hokiegeek2** [comment]
@Ethan-DeBandi99 @pierce314159 @joshmarshall1 @jeichert60 pulled in upstream master, ready for final review

**Comment by hokiegeek2** [comment]
@pierce314159 @Ethan-DeBandi99 @joshmarshall1 @jeichert60 ready for final review and merge

**Review by Ethan-DeBandi99 — APPROVED** [review]
Looks good to me

**Review by pierce314159 — APPROVED** [review]

**Pull request merged**